Major refactor

elkcert opened the pull request:
Recommendation is to review the PR by going through individual commits (they're _mostly_ self-contained) and making sure the program produces correct outputs.

Major changes:
- Rehauled the way we respond to events (read description for https://github.com/magley/my-pi-home/pull/4/commits/e0c475e87f683bb746e5a696690d480858d545ff)
- Restructured codebase so it's logically more separated
- Changed the way we handle concurrent printing (separate thread with a queue instead of scattered locks everywhere)

A couple of comments / minor changes:
- `PrintThread` was added for the purpose of fixing an odd edgecase introduced by making the threads daemons. Namely, if we started the CLI program and ran the listen command, then exited the program (CTRL+C twice), we would get an `EOFException` that I assume has something to do with the daemons getting blocked while putting items in the queue for reading. The fix for this is commented in `cli.py`. Funnily enough this fix doesn't necessitate the introduction of `PrintThread`, but I feel it's probably still a more elegant solution than putting locks everywhere. Also has the benefit of letting us queue several actuator events before we decide to go back to listening in the CLI (previously this didn't work due to silly locking behavior _I think_).
- `functools.partial` has been replaced with lambdas so the inferred return type of the `x_get_reader` functions is tidier
- The `components` module has been removed as I felt it was spreading the codebase too thin. We had too many files containing only two or three functions. Instead all the functions are now together in  `devices` module.

elkcert committed:
refactor: Simplify mbkp

elkcert committed:
feat: Crudely rework architecture

We now have an event thread that alone handles the various possible event states. This allows us to use wait() without any race conditions.
We also no longer create loop threads for devices which have no use for them. Additionally, the threads we do have were turned into daemon threads that stop executing the moment all non-daemon threads stop executing. I believe this is fine because we don't have to do any fancy cleanup of resources when we exit the reader threads. If, however, we do eventually need fancy resource cleanup, reintroducing the stop event shouldn't be a problem.

elkcert committed:
refactor: Replace MyPiEvent pin with sensor

elkcert committed:
fix: lambda reference nonsense

elkcert committed:
refactor: Miscellaneous

elkcert committed:
feat: Select interface type from command line

elkcert committed:
refactor: Cleaner CLI

elkcert committed:
refactor: Move config functions to common

elkcert committed:
refactor: Separate major logical units out of main

elkcert committed:
refactor: Move top-level modules to util

elkcert committed:
refactor: Consolidate device functions

elkcert committed:
refactor: Don't call setup on every read

elkcert committed:
refactor: Print queue, elegant exit

magley commented on actuators/buzzer.py:
```diff
@@ -3,25 +3,20 @@
-    return (do_buzz, stop_buzz)
\ No newline at end of file
+def do_buzz_simulated():
+     pass
```
Too many indentation spaces.

elkcert replied:
https://github.com/magley/my-pi-home/pull/4/commits/bf0ddd5d7d8225d6083945886ff7531babfc193a

elkcert committed:
refactor: Extract GUI and CLI

magley commented on common.py:
Was there a reason to move `SensorConfig` in the `common` module? What I would have done is kept it in `config` and rename `common` into `event`.

Even though we use `common` in most of our projects, I think it's a bad practice because the common module/package/class/namespace/... becomes cluttered with unrelated things.

Instead of `util/common`, `util/config`, `util/app`, why not `common/config`, `common/event`, `common/app`. When you think about it, the `App` module is essential to our program (`main.py` of all things imports it directly), so it's a little bit higher than "utility".

This also applies to `PrintThread`. The idea is that different "things" should be in separate files _unless_ they are tightly coupled (by logic, not by code and who-uses-who), so like `MyPiEventType` and `MyPiEvent`.

elkcert replied:
Agreed, much better this way. https://github.com/magley/my-pi-home/pull/4/commits/9c0f14479501dd3394739686a695282df12e5f8a

magley commented on util/app.py:
```diff
@@ -36,82 +35,23 @@ def run(self):
 
 
     def _run(self):
+        import ui.gui as gui
+        import ui.cli as cli
```
Why not import them inside the corresponding `case` below?

elkcert replied:
`cli.console_app` is called in two cases, would be redundant to put an import inside both of them. The reason why these aren't at the top level is circular imports.

magley reviewed (changes requested):
As a sort of sanity check, for all later devices the approach would be following:

- simulator functions for actuators are always empty
- for every new device type, aside from writing the sensor/actuator code, we also need to add it to `devices.py` with `event` functions for actuators or `setup + get reader` for sensors, as well as register a callback in `setup.py`
- new event handlers are written in `_event_thread` while the events are fired by the cli/gui
- everything else is handled by `App` and other modules

If the above is correct, I have a couple of questions:

1) What would be the preferred way of writing different callbacks for the same type of device (e.g. different behavior for PIR1 and PIR2's `on_motion`)?

2) How do I know if I want to set `ignore_paused` to `True` when putting a message in the print queue?

Aside from this and the comments I've left, everything else is fine.

---

>PrintThread was added for [...]

My clunky lock-based solution from the last PR was because of the `EOFException` which I couldn't catch for some reason. This is better, I completely forgot about daemon threads.

>The components module has been removed [...]

This is also great, it was confusing to have two `pir.py` open in the editor and having to navigate between a device and its component all the time.

magley commented on ui/cli.py:
```diff
@@ -0,0 +1,52 @@
+    print('Enter command:', end='')
+    i = input()
+
+    if i == 'room-buzz-on':
+        app.room_buzzer_on()
+    elif i == 'room-buzz-off':
+        app.room_buzzer_off()
+    elif i == 'door-light-on':
+        app.door_light_on()
+    elif i == 'door-light-off':
+        app.door_light_off()
+    elif i == 'listen':
+        app.print_thread.set_unpaused()
+        try:
+            while True:
+                time.sleep(1)
+        except KeyboardInterrupt:
+            print("Stopped listening...")
+    elif i == 'quit':
+        return True
+    else:
+        print("Unknown command")
```
Since you've used `match/case` elsewhere in this PR, this would be a good place to utilize it too?

elkcert replied:
It is a little nicer than if statements https://github.com/magley/my-pi-home/pull/4/commits/bf0ddd5d7d8225d6083945886ff7531babfc193a
Be aware though that match in python is a little wonky (for example, we couldn't extract the command strings to constants and match on those without some hackiness). Also, they're >=3.10 only.

magley replied:
Hopefully the RPis have >=3.10 installed, but that's something we should've checked last time...

elkcert replied:
Another "cutting-edge" python feature we're using are generic built-in types (for example list[int] instead of typing.List[int]). This is a >=3.9 feature.
A quick google search suggests the default python version on RasPis is 3.9...
Are we going to demonstrate the project on the RasPi for the first control point? That would be problematic also because we haven't actually tested the unsimulated branches of the code.

elkcert committed:
fix: Pesky queue pausing bug

In CLI, if we listen -> Ctrl+C -> <any command>, the command will print immediately because the print thread was blocking on self.queue.get instead of self.is_unpaused.wait().

elkcert committed:
Refactor: Match cli, fix indentation

elkcert committed:
refactor: util -> common, spread out modules

elkcert commented:
> ...If the above is correct

That's pretty much it.

> What would be the preferred way of writing different callbacks for the same type of device (e.g. different behavior for PIR1 and PIR2's on_motion)?

In `setup_devices()` in the appropriate `case` we would have another `match` on `cfg.name` and assign appropriate callbacks that way.

> How do I know if I want to set ignore_paused to True when putting a message in the print queue?

`ignore_paused` is only used when we want to buffer the queue while `print_thread` is paused. More specifically, we want to do this when we trigger actuator events that we want to see displayed once we unpause `print_thread` (all other `print_thread` puts are swallowed up). You can test it out in the CLI by triggering an event and then calling listen, (all) the triggered events will be displayed with the correct timestamp.
